fix-exception-handling

elekTom opened the pull request:
This PR aims to fix the exception handling interactions between Lua and C#.

At the moment C#-Lua functions just call `lua_error` from wherever the fault arises.
Since Lua uses `setjmp/longjmp` this means that a jump over a .NET managed stack will occur.
This somewhat works on Windows thanks to SEH. But only a [best effort basis](https://github.com/dotnet/runtime/pull/111259#issuecomment-2603732505), which is why it's currently [broken in NET9](https://github.com/dotnet/runtime/issues/111242).
On Linux the is no exception ABI comparable to SEH on Windows. Therefore jumping over managed frames won't work at all.
Due to these platform dependent incompatibilities `longjmp` is [officially unsupported](https://learn.microsoft.com/en-us/dotnet/standard/native-interop/exceptions-interoperability) by .NET.

Thus a more robust solution should be employed.
Effectively any and all exceptions have to be caught by the environment that raised them in the first place and forwarded in a stable fashion.

Commit:
Incorporate changes from ceac1a5f @ upstream

Commit:
Incorporate changes from a32b3b5d @ upstream

elekTom commented:
The current changes apply 2 commit made in upstream to improve the situation on Linux somewhat.
These changes basically make sure that `lua_error` is only ever called from the entry stackframe in the hopes that the jumped over stack will be minimal to none.
But since `longjmp` on Windows participates in the SEH .NET9  still notice it and raises a `SEHException` in response.

BTW: Besides the mentioned mitigations upstream doesn't seem to have fix the root problem yet either.

Commit:
Move build logic into build-reuse.yml and test on multiple platforms from there

Commit:
Some  fixes:

- use Buerkert.KeraLua version 1.7.0-14
- wait with build.pack until build completed

Commit:
build and test for both .NET 8 and 9

Commit:
add net9.0 as TargetFramework along with net8.0

Commit:
unset gh action `fail-fast`

Commit:
fix net version matrix build/test

theKBro commented:
But this also means that our apps using this library and throwing exceptions in registered functions will not work under linux?

Commit:
apply changes from https://github.com/NLua/NLua/pull/478

This should properly handle all C# exceptions by calling `lua_error` from a Lua wrapper, meaning no `longjmp` over managed frames will be performed

elekTom commented:
> But this also means that our apps using this library and throwing exceptions in registered functions will not work under linux?

As I see/plan it this should not interfere with any existing functionality.
Unless there is some behaviour I'm missing and that is also not covered in the unit tests.

elekTom commented:
The last commit 7acd0ba1 is intended to actually fix the exception handling by wrapping all C# functions in a Lua function that will call `lua_error` in case of an exception, meaning no `longjmp` over managed frames will be performed.

This implementation was taken from this [NLua-PR](https://github.com/NLua/NLua/pull/478), which seems to be abandoned for some unknown reason.

But as one can see now all test pass on both Linux and Windows running .NET8 and .NET9

Commit:
reduce to one deploy workflow that uses build-reuse internally

Commit:
fix assertion order in String ctype test

Commit:
rename build-deploy to deploy and add push trigger for master branch

Commit:
prepare for new version:

- update copyright year to 2026
- bump version to 1.8.0

Commit:
Handle unlikely pcall error in `PushWrappedCFunction`

Commit:
improve unit tests:

 - remove commented out code
 - use "NLuaTest.Mock.TestClass" instead of "NLuaTest.TestTypes.TestClass"

elekTom commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,40 @@
+    with:
+      version_suffix: ${{ github.event_name == 'release' && '' || github.run_number }}
+
+  deploy:
```
thats already there

Commit:
drop net9.0 support:

- update projects to target only net8.0
- adjust GitHub Actions workflow for single .NET version

elekTom commented:
@theKBro I dropped NET9 support again. Although it works now, I figure we will just skip NET9 all together.

theKBro commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,40 @@
+  build:
+    uses: ./.github/workflows/build-reuse.yml
+    with:
+      version_suffix: ${{ github.event_name == 'release' && '' || github.run_number }}
```
so sorry for the question but what exactly is this doing? 
on release it is '' else the run_number? 
this syntax looks a bit strange for me (with && and || )

theKBro replied:
from my point of view
```suggestion
      version_suffix: ${{ github.event_name == 'release' ? '' : github.run_number }}
```
is easier to read

elekTom replied:
Yeah, I'm not a fan either. Sadly Github actions doesn't have a ternary, [see issue here](https://github.com/actions/runner/issues/409#issuecomment-752775072).
So what this does, is to apply the run number to the build as long as it's not for release. As far as I can tell this was  also the intended behavior before.

elekTom replied:
Ok, there still was a problem. Since `''` is falsy  it will always result in `github.run_number`.
This is no fixed by inverting the condition.

theKBro commented on NLua/NLua.csproj:
```diff
@@ -34,7 +34,7 @@
     </ItemGroup>
     <ItemGroup>
-        <PackageReference Include="Buerkert.KeraLua" Version="1.6.0" />
+        <PackageReference Include="Buerkert.KeraLua" Version="1.7.0-14" /> <!-- TODO: update to 1.7.0 as soon as its available -->
```
todo

elekTom replied:
done

theKBro commented:
@elekTom could you also change the test dependencies to use wildcards?

Commit:
test gh action var eval

Commit:
rename test_var to version_suffix in build workflow

Commit:
test gh action var eval

Commit:
invert version_suffix condition for deploy workflow since `''` is falsy

Commit:
trigger release workflow on `released` since `published` will also trigger on drafts and pre-releases

elekTom commented:
> @elekTom could you also change the test dependencies to use wildcards?

@theKBro as far as I can tell, all test dependencies already use wildcards. But I could still update those.
Or should we do this through dependabot, which is paused at the moment.

theKBro commented:
> > @elekTom could you also change the test dependencies to use wildcards?
> 
> @theKBro as far as I can tell, all test dependencies already use wildcards. But I could still update those. Or should we do this through dependabot, which is paused at the moment.

sry I had a look at the wrong files -> forget about it

Commit:
update `Buerkert.KeraLua` to new v1.7.0 release

Commit:
bump test dependencies:

- update `Microsoft.NET.Test.Sdk` to v18.*
- update `NUnit3TestAdapter` to v6.*

elekTom commented:
@theKBro I now bumped the test dependencies
